src/front/wgsl/number.rs

```diff
@@ -0,0 +1,506 @@
+enum Kind {
+    Int(IntKind),
+    Float(FloatKind),
+}
+
+enum IntKind {
+    I32,
+    U32,
+}
+
+enum FloatKind {
+    F32,
+    F16,
+}
+
+// The following regexes (from the WGSL spec) will be matched:
+
+// int_literal:
+// | / 0                                                                [iu]?   /
+// | / [1-9][0-9]*                                                      [iu]?   /
+// | / 0[xX][0-9a-fA-F]+                                                [iu]?   /
+
+// decimal_float_literal:
+// | / 0                                                                [fh]    /
+// | / [1-9][0-9]*                                                      [fh]    /
+// | / [0-9]*               \.[0-9]+            ([eE][+-]?[0-9]+)?      [fh]?   /
+// | / [0-9]+               \.[0-9]*            ([eE][+-]?[0-9]+)?      [fh]?   /
+// | / [0-9]+                                    [eE][+-]?[0-9]+        [fh]?   /
+
+// hex_float_literal:
+// | / 0[xX][0-9a-fA-F]*    \.[0-9a-fA-F]+      ([pP][+-]?[0-9]+        [fh]?)? /
+// | / 0[xX][0-9a-fA-F]+    \.[0-9a-fA-F]*      ([pP][+-]?[0-9]+        [fh]?)? /
+// | / 0[xX][0-9a-fA-F]+                         [pP][+-]?[0-9]+        [fh]?   /
+
+// You could visualize the regex below via https://debuggex.com to get a rough idea what `extract_info` is doing
+// -?(?:0[xX](?:([0-9a-fA-F]+\.[0-9a-fA-F]*|[0-9a-fA-F]*\.[0-9a-fA-F]+)(?:([pP][+-]?[0-9]+)([fh]?))?|([0-9a-fA-F]+)([pP][+-]?[0-9]+)([fh]?)|([0-9a-fA-F]+)([iu]?))|((?:[0-9]+[eE][+-]?[0-9]+|(?:[0-9]+\.[0-9]*|[0-9]*\.[0-9]+)(?:[eE][+-]?[0-9]+)?))([fh]?)|((?:[0-9]|[1-9][0-9]+))([iufh]?))
+
+fn extract_info(input: &str) -> Option<ExtractInfo<'_>> {
+    macro_rules! consume {
+        ($bytes:ident, $($($pattern:pat)|*),*) => {
+            match $bytes {
+                &[$($($pattern)|*),*, ref rest @ ..] => (rest, true),
+                bytes => (bytes, false),
+            }
+        };
+    }
+
+    macro_rules! consume_map {
+        ($bytes:ident, [$($($pattern:pat)|* => $to:expr),*]) => {
+            match $bytes {
+                $( &[$($pattern)|*, ref rest @ ..] => (rest, Some($to)), )*
+                bytes => (bytes, None),
+            }
+        };
+    }
+
+    const fn consume_dec_digits(input: &[u8]) -> (&[u8], usize) {
+        let mut output = input;
+        while let &[b'0'..=b'9', ref rest @ ..] = output {
+            output = rest;
+        }
+        (output, input.len() - output.len())
+    }
+
+    const fn consume_hex_digits(input: &[u8]) -> (&[u8], usize) {
+        let mut output = input;
+        while let &[b'0'..=b'9' | b'a'..=b'f' | b'A'..=b'F', ref rest @ ..] = output {
+            output = rest;
+        }
+        (output, input.len() - output.len())
+    }
+
+    const fn to_str(bytes: &[u8]) -> &str {
+        // SAFETY: the bytes are valid UTF-8 since the buffer
+        // was created from a &str and we only advance
+        // it by one byte codepoints
+        unsafe { core::str::from_utf8_unchecked(bytes) }
+    }
```

Comment:
Instead of the unsafe code, please use the length of `bytes` versus the length of `input` in bytes to do a fresh slice of `input`.